### PR TITLE
Partest: add `instrumented` test category.

### PR DESCRIPTION
**Pull request by gkossakowski**
---

(taken from README)

Tests in `instrumented/` directory are executed the same way as in `run/` but
they have additional byte-code instrumentation performed for profiling. You
should put your tests in `instrumented/` directory if you are interested in
method call counts. Examples include tests for specialization (you want to
count boxing and unboxing method calls) or high-level tests for optimizer
where you are interested if methods are successfuly inlined (so they should
not be called at runtime) or closures are eliminated (so no constructors
of closures are called).

Check `scala.tools.partest.instrumented.Instrumentation` to learn how to
use the instrumentation infrastructure.

The instrumentation itself is achieved by attaching a Java agent to the
forked VM process that injects calls to profiler. Check
`scala.tools.partest.javaagent.ProfilingAgent` for details.

---

A few notes on low-level details of this change:
- Partest now depends on asm library for byte-code instrumentation
  (`build.xml`)
- Build additional jar called `scala-partest-javaagent.jar` that is
  used with `-javaagent:` option. (`build.xml`)
- Set `-javaagent:` option for all tests in `instrumented/` directory.
  (`RunnerManger.scala`)
- Introduce a new category of tests called `instrumented`.
- Add one instrumented test to demonstrate usage and test new
  infrastructure itself. (`InstrumentationTest.scala`)

Review by @phaller.

### CODE REVIEW

**Comment by gkossakowski** [comment]
I think @alex22, @VladUreche, @magarciaEPFL might be intested to have a look at this infrastructure piece because it's specialization/optimizer related.

Also mentioning @paulp and @jsuereth that are generally interested in partest I think.

File: src/partest/scala/tools/partest/instrumented/Intrumentation.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ * should print:
+ *
+ * {{{
+ * true
+ * Method call statistics:
+ * scala/Predef$.println(Ljava/lang/Object;)V: 1
+ * scala/runtime/BoxesRunTime.boxToBoolean(Z)Ljava/lang/Boolean;: 1
+ * }}}
+ */
+object Instrumentation {
```
Comment (phaller) [inline src/partest/scala/tools/partest/instrumented/Intrumentation.scala]: Btw, there seems to be a typo in the file name ("Intrumentation.scala").

**Comment by phaller** [comment]
LGTM

**Comment by adriaanm** [comment]
Note that this pull request is against master. I thought @gkossakowski wanted to have this for 2.10.
I'm going to hold off merging until we can clarify. (Clean history above everything!)

**Comment by gkossakowski** [comment]
@adriaanm: You are right. I wanted this to go into 2.10.x branch. I'll create a new pull request against a correct branch. Also, I'll fix the typo @phaller pointed out.

**Pull request closed without merging**